src/main/java/com/fasterxml/jackson/databind/jsontype/impl/TypeNameIdResolver.java

```diff
@@ -80,7 +82,16 @@ public static TypeNameIdResolver construct(MapperConfig<?> config, JavaType base
     @Override
     public String idFromValue(Object value)
     {
-        Class<?> cls = _typeFactory.constructType(value.getClass()).getRawClass();
+       return idFromClass(value.getClass());
+    }
+
+
+    protected String idFromClass(Class<?> clazz)
+    {
+        if(clazz==null){
```

Comment:
Fair point albeit it is backward compatible :-) Previous impl:
```java
    @Override
    public String idFromValue(Object value)
    {
        Class<?> cls = _typeFactory.constructType(value.getClass()).getRawClass();
```

Please let me know if you want this changed.